linera-storage/src/db_storage.rs

```diff
@@ -195,50 +204,124 @@ enum BaseKey {
 }
 
 /// A clock that can be used to get the current `Timestamp`.
+#[async_trait]
 pub trait Clock {
     fn current_time(&self) -> Timestamp;
+
+    async fn sleep(&self, delta: TimeDelta);
+
+    async fn sleep_until(&self, timestamp: Timestamp);
 }
 
 /// A `Clock` implementation using the system clock.
 #[derive(Clone)]
 pub struct WallClock;
 
+#[async_trait]
 impl Clock for WallClock {
     fn current_time(&self) -> Timestamp {
         Timestamp::now()
     }
+
+    async fn sleep(&self, delta: TimeDelta) {
+        tokio::time::sleep(delta.as_duration()).await
+    }
+
+    async fn sleep_until(&self, timestamp: Timestamp) {
+        let delta = timestamp.delta_since(Timestamp::now());
+        if delta > TimeDelta::ZERO {
+            self.sleep(delta).await
+        }
+    }
+}
+
+#[cfg(with_testing)]
+#[derive(Default)]
+struct TestClockInner {
+    time: Timestamp,
+    sleeps: BTreeMap<Reverse<Timestamp>, Vec<oneshot::Sender<()>>>,
+}
+
+#[cfg(with_testing)]
+impl TestClockInner {
+    fn set(&mut self, time: Timestamp) {
+        self.time = time;
+        let senders = self.sleeps.split_off(&Reverse(time));
+        for sender in senders.into_values().flatten() {
+            let _ = sender.send(());
+        }
+    }
+
+    fn add_sleep(&mut self, delta: TimeDelta) -> Receiver<()> {
+        let (sender, receiver) = oneshot::channel();
+        if delta == TimeDelta::ZERO {
+            let _ = sender.send(());
+        } else {
+            let until = self.time.saturating_add(delta);
+            self.sleeps.entry(Reverse(until)).or_default().push(sender);
+        }
+        receiver
```

Comment:
Optional: maybe reuse `add_sleep_until` here?

```suggestion
        self.add_sleep_until(self.time.saturating_add(delta))
```